OSSMDOC-205: Document configuring federating meshes. (#35580)
openshift/openshift-docs: modules/ossm-federation-config-smcp.adoc

////
This module included in the following assemblies:
* service_mesh/v2x/ossm-federation.adoc
////

[id="ossm-federation-config-smcp_{context}"]
= Configuring a control plane for federation

> Before a mesh can be federated, you must configure the `ServiceMeshControlPlane` for mesh federation.   Because  all meshes that are members of the federation are equal, and each mesh is managed independently,  you must configure the SMCP for _each_ mesh that will participate in the federation.

Reviewer: ```suggestion
Before a mesh can be federated, you must configure the `ServiceMeshControlPlane` for mesh federation. Because  all meshes that are members of the federation are equal, and each mesh is managed independently, you must configure the SMCP for _each_ mesh that will participate in the federation.
```
Surplus spaces?

In the following example, the administrator for the `red-mesh` is configuring the SMCP for federation with both the `green-mesh` and the `blue-mesh`.

.Sample SMCP for red-mesh
[source,yaml]
----
apiVersion: maistra.io/v2
kind: ServiceMeshControlPlane
metadata:
  name: red-mesh

>   namespace: red-istio-system

Reviewer 2: In the previous example, the namespace was `red-mesh-system`

```suggestion
  namespace: red-mesh-system
```

spec:
  version: v2.1
  cluster:
    name: red-cluster

>     network: red-network

Reviewer 3: @rcernich do we need to mention .spec.cluster.network in federation SMCP CR ?

Reviewer 4: Neither of these is required.

Author: So should lines 24 and 25 be deleted?

Reviewer 5: yep. the whole cluster block (l23-l25) can be removed

Author: deleted

|

|spec:
  gateways:
    additionalEgress:
      <egressName>:
        router mode:
|
|`sni-dnat`
|

|spec:
  gateways:
    additionalEgress:
      <egressName>:
        service:
          metadata:
            labels:
              federation.maistra.io/proxy:
|Specify a unique label for the gateway to prevent federated traffic from flowing through the cluster's default system gateways.
|
|

|spec:
  gateways:
    additionalEgress:
      <egressName>:
        service:
          ports:
|Used to specify the `port:` and `name:` used for TLS and service discovery.  Federation traffic consists of raw encrypted TCP for service traffic.
|Port `15443` is required for sending TLS service requests to other meshes in the federation. Port `8188` is required for sending service discovery requests to other meshes in the federation.
|

|spec:
  gateways:
    additionalIngress:
|Define an additional ingress gateway gateway for _each_ mesh peer in the federation.
|
|

|spec:
  gateways:

>     additionalEgress:

Reviewer 6: Hi Julie,

The parameter is called additionalEgress and the description mentions ingress.

Author: Thanks, good catch.  Copy/paste error!

      <ingressName>:
        enabled:
|This parameter enables or disables the federation ingress.
|`true`/`false`
|`true`

|spec:
  gateways:
    additionalIngress:
      <ingressName>:
        router mode:
|
|`sni-dnat`
|

|spec:
  gateways:
    additionalIngress:
      <ingressName>:
        service:
          type:

> |The ingress gateway service must be exposed through a load balancer that operates at Layer4 of the OSI model and is publicly available.

Reviewer: ```suggestion
|The ingress gateway service must be exposed through a load balancer that operates at Layer 4 of the OSI model and is publicly available.
```

|===

== Understanding federation trust domain parameters

Each mesh in the federation must have its own unique trust domain. This value is used when configuring mesh federation in the `ServiceMeshPeer` resource.

[source,yaml]
----
kind: ServiceMeshControlPlane
metadata:
  name: red-mesh
  namespace: red-mesh-system
spec:
  security:
    trust:
      domain: red-mesh.local
----

.Federation security parameters
[options="header"]
[cols="l, a, a, a"]
|===
|Parameter |Description |Values |Default value
|spec:
  security:
    trust:
      domain:
|Used to specify a unique name for the trust domain for the mesh. Domains must be unique for every mesh in the federation.
|`<mesh-name>.local`
|N/A
|===

== Understanding federation DNS parameters

The DNS parameters are configured automatically in the version 2.1 SMCP; you do not need to configure them. They are presented here for reference purposes.

.Federation DNS parameters
[source,yaml]
----
kind: ServiceMeshControlPlane
metadata:
  name: red-mesh
  namespace: red-mesh-system
spec:
  techPreview:
    meshConfig:
      defaultConfig:
        proxyMetadata:
          ISTIO_META_DNS_CAPTURE: "true"
          PROXY_XDS_VIA_AGENT: "true"

> ----

Reviewer 4: All of the above `techPreview` settings are not required and are configured by default.

Author: Yes, and later in the file I point out that those lines are configured by default.  Question, if they're not required, then it won't matter if customers deleted those lines?  My worry was that if we didn't document them, that someone might delete that configuration and it might break something.

Reviewer 4: They are required and the only way they can be disabled is if they add the block above, but change the settings to false.  I consider these internal details, which users shouldn't have to worry about.  All they need to know is that DNS requests are managed by the proxy.

Author: As long as you think customers can't hurt themselves, then should I remove this entire heading and section?

Reviewer 4: Yes, delete the whole DNS section.

Author: Deleted DNS section.

.Federation DNS parameters
[options="header"]
[cols="l, a, a, a"]
|===
|Parameter |Description |Values |Default value
|ISTIO_META_DNS_CAPTURE:
|This setting is required so that the proxy handles DNS requests, which are delegated to Istiod.
|`true`/`false`
|`true`

|PROXY_XDS_VIA_AGENT:
|This setting is required to enable DNS capture.
|`true`/`false`
|`true`
|===

> 

Reviewer 4: The above settings for DNS are not required, as they're configured automatically.

Author: A couple of lines up "They are presented here for reference purposes."

Reviewer 4: I don't think we want to document these.  If folks have issues because DNS proxy doesn't work correctly, we would given them these as a workaround.  In practice, we don't want users changing these values.

Reviewer 4: And you can delete this DNS section too.

. On the *Create ServiceMeshControlPlane Details* page, click `YAML` to modify your configuration.

. Modify your `ServiceMeshControlPlane` to add federation ingress and egress gateways and to specify the trust domain.

. Click *Save*.

.Procedure from the CLI

Follow this procedure to create or edit the `ServiceMeshControlPlane` with the command line. This example uses the `red-mesh` as an example.

. Log in to the {product-title} CLI as a user with the `cluster-admin` role. Enter the following command. Then, enter your username and password when prompted.
+
[source,terminal]
----
$ oc login --username=NAMEOFUSER https://{HOSTNAME}:6443
----
+
. Change to the project where you installed the control plane, for example red-mesh-system.
+
[source,terminal]
----
$ oc project red-mesh-system
----
+
. Edit the `ServiceMeshControlPlane` file to add federation ingress and egress gateways and to specify the trust domain.

. Run the following command to edit the control plane where `red-mesh-system` is the system namespace and `red-mesh-install.yaml` includes a full path to the file you edited:
+
[source,terminal]
----
$ oc edit -n red-mesh-system -f red-mesh-install.yaml
----
+

> . Enter the following command,where `red-mesh-system` is the system namespace, to see the status of the control plane installation.

Reviewer 5: missing a whitespace after `command,`

> . Execute the following command,where `red-mesh-system` is the system namespace, to see the status of the control plane installation.

Reviewer: ```suggestion
. Enter the following command, where `red-mesh-system` is the system namespace, to see the status of the control plane installation.
```
IBM SG "execute: Use only if a simpler term, such as "run", "issue", "start", or "enter" is inappropriate."
